app/ide-desktop/lib/dashboard/src/authentication/src/http.tsx

```diff
@@ -0,0 +1,178 @@
+    json = (body: Json) => {
+        this.request.body = JSON.stringify(body);
+        this.request.headers.set("Content-Type", "application/json");
+        return this;
+    };
+
+    /** Finalizes the HTTP request and sends it using the HTTP client. */
+    send = () => this.client.executeRequest(this.request);
+}
+
+// === RequestInitWithUrlAndHeaders ===
+
+/** An HTTP request that is guaranteed to also have the URL it will be send to and an initialized map
+ * of {@link Headers}.
+ *
+ * This is an extension of the built-in {@link RequestInit} interface, which does not provide the
+ * aforementioned guarantees. In fact, it does not provide a `url` at all, and the `headers` field
+ * may even be `undefined` for that interface. */
+interface RequestInitWithUrlAndHeaders extends RequestInit {
+    url: string;
+    headers: Headers;
+}
+
+
+
+// ==============
+// === Client ===
+// ==============
+
+/** An HTTP client that can be used to create and send HTTP requests asynchronously. */
+export class Client {
+    /** A map of default headers that are included in every HTTP request sent by this client.
+     *
+     * This is useful for setting headers that are required for every request, like authentication
+     * tokens. */
+    defaultHeaders?: Headers;
+
+    /** Returns a new {@link ClientBuilder} that can be used to create an HTTP {@link Client}. */
+    static builder = () => new ClientBuilder();
+
+    /** Returns a builder for an HTTP GET request to the specified URL. */
+    get = (url: string) => this.request(HttpMethod.get, url);
+
+    /** Returns a builder for an HTTP POST request to the specified URL. */
+    post = (url: string) => this.request(HttpMethod.post, url);
+
+    /** Returns a builder for an HTTP request to the specified URL, with the given HTTP method. */
+    request = (method: HttpMethod, url: string) => {
+        const config = { method };
+        return new RequestBuilder(this, url, config);
+    };
+
+    /** Asynchronously sends an HTTP request, returning a wrapped {@link HttpResponse}, which can be
+     * used to access the status code and deserialize the response body. */
+    executeRequest = async (config: RequestInitWithUrlAndHeaders) => {
+        /** Insert default headers into the request headers without overwriting already appended
+         * headers. */
+        if (this.defaultHeaders != null) {
+            this.defaultHeaders.forEach((value, key) => {
+                if (!config.headers.has(key)) {
+                    config.headers.set(key, value);
+                }
+            });
+        }
+
+        /** Construct the request from the options. */
+        const request = new Request(config.url, config);
+
+        const response = await fetch(request);
+
+        return new HttpResponse(response);
+    };
+}
+
+
+
+// =====================
+// === ClientBuilder ===
+// =====================
+
+/** Builder for a new HTTP {@link Client}. */
+class ClientBuilder {
```

Comment:
why not just add a constructor to `Client`?

Reply:
The `Client` was more complex when the builder was initially written. It was simplified so the builder is now a bit redundant. I'll remove it.